zkevm-circuits/src/pi_circuit2.rs

```diff
@@ -380,1147 +220,479 @@ impl<F: Field> SubCircuitConfig<F> for PiCircuitConfig<F> {
-        // at expected offset
-        meta.create_gate("tx_table.tx_id[i] == rpi[i]", |meta| {
-            // row.q_tx_table * row.tx_table.tx_id
-            // == row.q_tx_table * row_offset_tx_table_tx_id.raw_public_inputs
-            let q_tx_table = meta.query_selector(q_tx_table);
-            let tx_id = meta.query_advice(tx_table.tx_id, Rotation::cur());
-            let rpi_tx_id = meta.query_advice(rpi, Rotation(offset as i32));
-
-            vec![q_tx_table * (tx_id - rpi_tx_id)]
-        });
+        // field bytes
+        meta.create_gate(
+            "rpi_field_bytes_acc[i+1] = rpi_field_bytes_acc[i] * t + rpi_bytes[i+1]",
+            |meta| {
+                let mut cb = BaseConstraintBuilder::new(MAX_DEGREE);
+
+                let q_field_step = meta.query_selector(q_field_step);
+                let rpi_field_bytes_acc_next =
+                    meta.query_advice(rpi_field_bytes_acc, Rotation::next());
+                let rpi_field_bytes_acc = meta.query_advice(rpi_field_bytes_acc, Rotation::cur());
+                let rpi_field_bytes_next = meta.query_advice(rpi_field_bytes, Rotation::next());
+                let is_field_rlc = meta.query_fixed(is_field_rlc, Rotation::next());
+                let randomness = challenges.evm_word();
+                let t = select::expr(is_field_rlc, randomness, BYTE_POW_BASE.expr());
+                cb.require_equal(
+                    "rpi_field_bytes_acc[i+1] = rpi_field_bytes_acc[i] * t + rpi_bytes[i+1]",
+                    rpi_field_bytes_acc_next,
+                    rpi_field_bytes_acc * t + rpi_field_bytes_next,
+                );
+                cb.gate(q_field_step)
+            },
+        );
+        meta.create_gate("rpi_field_bytes_acc[0] = rpi_field_bytes[0]", |meta| {
+            let mut cb = BaseConstraintBuilder::new(MAX_DEGREE);
 
-        meta.create_gate("tx_table.index[i] == rpi[tx_table_len + i]", |meta| {
-            // row.q_tx_table * row.tx_table.tx_index
-            // == row.q_tx_table * row_offset_tx_table_tx_index.raw_public_inputs
-            let q_tx_table = meta.query_selector(q_tx_table);
-            let tx_index = meta.query_advice(tx_table.index, Rotation::cur());
-            let rpi_tx_index = meta.query_advice(rpi, Rotation((offset + tx_table_len) as i32));
+            let q_field_start = meta.query_selector(q_field_start);
+            let rpi_field_bytes_acc = meta.query_advice(rpi_field_bytes_acc, Rotation::cur());
+            let rpi_field_bytes = meta.query_advice(rpi_field_bytes, Rotation::cur());
 
-            vec![q_tx_table * (tx_index - rpi_tx_index)]
+            cb.require_equal(
+                "rpi_field_bytes_acc[0] = rpi_field_bytes[0]",
+                rpi_field_bytes_acc,
+                rpi_field_bytes,
+            );
+            cb.gate(q_field_start)
         });
+        meta.create_gate("rpi_field_bytes_acc[last] = rpi", |meta| {
```

Comment:
Is this copy necessary?

Reply:
You mean we need  rpi_field_bytes_acc ==  instance directly?

Reply:
Sure why not? I think it's possible the `rpi` column could be removed entirely, the use of `rpi_rlc_acc`/`rpi` also confuses me a bit, seems easier not to use `rpi` to me.